### DISCUSSION TITLE
Cron jobs

### DISCUSSION THREAD

**Discussion by iammrsea** [adonisjs/core#1277] [Help] [answered]
Hello guys, I would appreciate if anyone could assist me with this problem.  I have a cron job I would like to start after application boot, but I don't seem to have figured out the appropriate place to start it. I have tried the AppProvider lifecycle hooks, but it failed with 'E_IOC_LOOKUP_FAILED: Cannot resolve Adonis/Lucid/Orm namespace from the IoC Container'. Obviously, my cron job service provider depends on some of my Lucid models. Any ideas? Of course, I am using V5

**Comment by McSneaky** [comment, marked as answer]
Getting same error as in here: https://forum.adonisjs.com/t/adonis-5-command-database-ioc-error/6217
Maybe this solves it too :)

```ts
import { BaseCommand } from '@adonisjs/ace'
import { Kernel } from '@adonisjs/ace'
import { inject } from '@adonisjs/fold'
import { DatabaseContract } from '@ioc:Adonis/Lucid/Database'
import { ApplicationContract } from '@ioc:Adonis/Core/Application'

@inject([null, null, 'Adonis/Lucid/Database'])
export default class Hurr extends BaseCommand {
  public static commandName = 'hurr'
  public static description = 'Durr'

  constructor (app: ApplicationContract, kernel: Kernel, private db: DatabaseContract) {
    super(app, kernel)
  }

  public static settings = {
    loadApp: true,
  }

  public async handle () {
    console.log(this.db)

    this.logger.info('Hello world!')
  }
}
```

**Reply by RomainLanz** [reply]
Yes, you need to tell `Ace` to load the full application to use them!

You can do that by adding the following code to your command:

```ts
public static settings = {
  loadApp: true,
}
```

**Reply by thetutlage** [reply]
What you guys think of writing the cron job logic outside of AdonisJS and that code just runs an ace command?

**Reply by RomainLanz** [reply]
I believe the name `cron` is wrongly used here since he wants to start it when application boot.
What he needs is a queue/scheduler.

**Reply by McSneaky** [reply]
Been using CRON calling Ace commands as scheduler, but it would be cool indeed to have something like that built in (or as community package)

**Reply by RomainLanz** [reply]
I saw that @reg2005 [made a package](https://www.npmjs.com/adonis5-scheduler) to schedule tasks.

**Reply by iammrsea** [reply]
> Getting same error as in here: https://forum.adonisjs.com/t/adonis-5-command-database-ioc-error/6217
> Maybe this solves it too :)
> 
> ```ts
> import { BaseCommand } from '@adonisjs/ace'
> import { Kernel } from '@adonisjs/ace'
> import { inject } from '@adonisjs/fold'
> import { DatabaseContract } from '@ioc:Adonis/Lucid/Database'
> import { ApplicationContract } from '@ioc:Adonis/Core/Application'
> 
> @inject([null, null, 'Adonis/Lucid/Database'])
> export default class Hurr extends BaseCommand {
>   public static commandName = 'hurr'
>   public static description = 'Durr'
> 
>   constructor (app: ApplicationContract, kernel: Kernel, private db: DatabaseContract) {
>     super(app, kernel)
>   }
> 
>   public static settings = {
>     loadApp: true,
>   }
> 
>   public async handle () {
>     console.log(this.db)
> 
>     this.logger.info('Hello world!')
>   }
> }
> ```

This was very helpful, thanks.

# My use case: 
## A client wants to be able to send health tips notification to his customers once per day at a fixed time, say 6am.

My working implementation:

```javascript
import NotificationService from "../services/NotificationService";
import { BaseCommand } from "@adonisjs/ace";
import { inject } from "@adonisjs/fold";
import { Kernel } from "@adonisjs/ace";
import { schedule } from "node-cron";
import { ApplicationContract } from "@ioc:Adonis/Core/Application";
import { DatabaseContract } from "@ioc:Adonis/Lucid/Database";
import FirebaseConfig from "../config/firebase";

@inject([null, null, "Adonis/Lucid/Database"])
export default class Notification extends BaseCommand {
  public static commandName = "notification";
  public static description = "Notification cron command";
  private readonly provider = new NotificationService(FirebaseConfig);

  constructor(
    app: ApplicationContract,
    kernel: Kernel,
    private db: DatabaseContract
  ) {
    super(app, kernel);
  }

  public static settings = {
    loadApp: true,
  };
  public async handle() {
   console.log('scheduling notifications...);
    await this.scheduleHealthTips("* 6 * * *");
  }
  private async scheduleHealthTips(cronExpression: string) {
    schedule(cronExpression, async () => {
      //my logic here
    });
  }
 
}
```
I run the serve and notification ace commands with npm-run-all package

**Reply by bitkidd** [reply]
Why not just use `crontab` and fire an ace command? I am not sure about schedulers written in js, crontab is much more low level and will fire anyway, even if server restarted or something happened with your app and no need to use any package, just run `crontab -e` and then you write something like this `* 6 * * * /path/to/your/app/node ace notification`

**Reply by AGDholo** [reply]
Laravel has same: https://laravel.com/docs/8.x/scheduling.  
If I want to run some simple job, I prefer to use were built into the scheduling framework,  

The advantage of doing this is that it is managed in one place, which is easy enough to migrate, and I only need one line of command on the server, so I don't feel overwhelmed after resetting the server system.
